### ISSUE TITLE
Wally does not hoist types to the linker module

### ISSUE THREAD

**Issue by vocksel** [UpliftGames/wally#85]
For instance, my package [ModuleLoader](https://github.com/vocksel/module-loader/) exports a top-level `Class` type which is simply `export type Class = typeof(ModuleLoader.new())`.

A problem arises in that Wally does not re-export this type in the generated `Packages/ModuleLoader.lua` file, which makes the type inaccessible to my code that consumes the package.

For example, the following fixes my issue and allows Luau analysis to work properly:

```lua
local package = require(script.Parent._Index["vocksel_module-loader@0.3.1"]["module-loader"])

export type Class = package.Class

return package
```

**Comment by LPGhatguy** [comment]
Is there syntax that Wally can use to export all types? It's probably not feasible for Wally to figure out what types are exported and generate the correct Luau to export them all.

**Comment by Corecii** [comment]
I'm not aware of any syntax to export *all* types, but I don't think it would be particularly hard for Wally to do on its own.

Types are *always* exported with `export type` statements, and can always be read as`module.TypeName`.

Reading a list of types would require parsing the root Luau file for `export type` statements. This may be unreasonably complex, though, and I'd personally be fine with a `types = ["TypeName"]` `wally.toml` property if it means we can use types. Any access to types is better than none!

Exporting them would be simple: `export type TypeName = package.TypeName` for every exported type.

**Comment by LPGhatguy** [comment]
> Reading a list of types would require parsing the root Luau file for export type statements.

This is one of the things that I think would be tricky. Because Wally packages can be anything, we don't know where the root Luau file is or if there even is one for sure! The worst case might be that you can publish an `rbxm` file as a package, which would be very tricky to inspect at package installation time.

**Comment by Uncontained0** [comment]
This would be a great feature to have, and while I think that parsing a `rbxm` file for types isn't in the scope of this issue, it would
be very beneficial if we were able to define an options `types` list in our `wally.toml` which would export types.

**Comment by Hexcede** [comment]
@LPGhatguy I know this is probably a little bit iffy, but, could you potentially use `ModuleName.project.json` instead of `ModuleName.lua`? As I understand it the module file exists to direct the specific name to the specific module and using `.project.json` files to point to the root project folder would achieve that.

I think it might also be valid to use a symlink to the root project folder (e.g. a relative symlink to `../_Index/author_package@version/package`).

Rojo seems to understand `default.project.json` in the same way as `init` files, and `.project.json` files can understandably be converted directly into instances like any other file, so either of these might be a valid way to go about it but I am not really sure.

**Comment by LPGhatguy** [comment]
> @LPGhatguy I know this is probably a little bit iffy, but, could you potentially use `ModuleName.project.json` instead of `ModuleName.lua`? As I understand it, the module file exists to direct the specific name to the specific version, right? Using `.project.json` files to point to the root project folder could also achieve that I think.
> 
> Problem then is that it would produce a lot of additional clutter, there would be duplicates of packages, which is not ideal.
> 
> I think it might also be valid to use a symlink to the root project folder (e.g. a relative symlink to `../_Index/author_package@version/package`).

Duplicates of packages are not just "not ideal", they fail the fundamental purpose of a package manager. Dependencies must be deduplicated and shared, we can't just copy them into every package that depends on them.

Wally already stores all of the installed packages in a single place. The linkage model between packages has been well thought-out and is not likely to change.

The best solution for this problem is for Luau to gain the ability to re-export types from another module. Wally's linkage modules could use that feature to great effect. Barring that, an entry in the manifest file listing types to re-export is reasonable.

**Comment by Hexcede** [comment]
Yeah, I realized why pretty much everything I said is flawed after the fact. After having looked into [that project](https://github.com/JohnnyMorganz/wally-package-types) that I was linked to, I think it solves the problem pretty well. It is also written in Rust. Perhaps the project could be integrated into Wally?

**Comment by magnalite** [comment]
I have given this some thought and read this discussion and come to some conclusions on how we will move forward here. I agree with the general sentiment and ideas you've all come up with.  I see two paths forward to enabling wally to support type exports.

---
### Manifest declared type exports
As mentioned before we can add a new field `types` (or perhaps `exported_types` to be more explicit?) to define an array of types exported by a package. This makes it very easy for the linker files to reexport the types. This also doesn't require any knowledge of the contents of a package, we can just reexport blind. It is also nice to act as a form of self documentation.

This is the immediate way we will move forward with supporting type exports. Primarily as this is a simple solution and is completely build tool agnostic.

---
### Rojo inspected type exports
If we move ahead with https://github.com/UpliftGames/wally/issues/142 and become rojo aware then we can get smarter (and more complex). If we identify a package as being "rojo-compatible" we can determine what the root of the package will be and inspect the script to discover type names. We can then auto re-export these types for the user without any further actions. We may want to prompt the user or automatically add them to the manifest. What is a big win here is all existing packages will suddenly have their types automagically exported and work.

We can add this at a later date as a new issue after declared types have been implemented.

**Comment by dev-syn** [comment]
> I have given this some thought, read this discussion, and come to some conclusions on how we will move forward here. Thanks for all your thoughts. I agree with the general sentiment and ideas you've all come up with. I see two paths forward to enabling wally to support type exports.
> 
> ### Manifest declared type exports
> As mentioned before we can add a new field `types` (or perhaps `type-exports` to be more explicit?) to define an array of types exported by a package. This makes it very easy for the linker files to re-export the types. This also doesn't require any knowledge of the contents of a package, we can just re-export blind. It is also nice to act as a form of self documentation.
> 
> This is the immediate way we will move forward with supporting type exports. Primarily due to this being a simple solution and build tool agnostic.
> 
> ### Rojo inspected type exports
> If we move ahead with #142 and become rojo aware then we can get smarter (and more complex). If we identify a package as being "rojo-compatible" we can determine what the root of the package will be and inspect the script to discover type names. We can then auto re-export these types for the user without any further actions. We may want to prompt the user or automatically add them to the manifest. What is a big win here is all existing packages will suddenly have their types automagically exported and work.
> 
> We can add this at a later date as a new issue after declared types have been implemented.

Hey, I'm just wondering if there is a general idea on when we will see our types be hoisted with the linker module?

This was said around June and we are now October and this feature is greatly needed for those who use types without having to use a third party tool

**Comment by Corecii** [comment]
Going to throw this out there:
If I were implementing this, here's what I'd do.
- declared type exports
- and basic inspected type exports supporting only the current version of Rojo

And if Rojo updates? We're not supporting its changes for inspected type exports. We can't keep up with every tool's format. The inspected type exports would be primarily for backwards-compatibility with the `wally-package-types` tool, and new code should explicitly export types. (Really, this shouldn't be our responsibility. This should be a Luau thing, but it's not so it's on us. 😞)

Additionally, _I_ wouldn't actually parse the Luau code. I would do a super basic string pattern search for `export type` definitions. If we happen to export a type that doesn't exist that's fine — all of the other types still work, and you shouldn't be looking for type errors in the Packages folders anyways. Here's a lune script along the lines of what I'm thinking: https://github.com/Corecii/lune-scripts/blob/main/lune/wally-install.lua

This keeps the dependencies light, the compatibility high, our responsibilities low, and implementation time low.

**Comment by Dekkonot** [comment]
I think parsing Luau is maybe a bit much but I do worry about blindly pattern matching just because I don't know how luau-analyze plays with broken type exports. If it works now, is that a promise it'll always work in the future? We could probably write a basic lexer and use that, though I'm not sure what it'd look like in practice. It might be worth exploring just in case, though it obviously has a cost in complexity and time.

With regards to Rojo's project format I don't see it changing significantly any time soon. We should be able parse it and read through it with no difficulties, especially since we only really care about what the module turns into when it's require.

I'm not sure this *should* actually be Luau's responsibility, incidentally. It seems like a rather niche use case to re-export every type from a module when the intended usage is for people to just... directly require modules and for modules to export types they want.

**Comment by Corecii** [comment]
> but I do worry about blindly pattern matching just because I don't know how luau-analyze plays with broken type exports. If it works now, is that a promise it'll always work in the future?

I would consider that a promise. If code with broken type exports works fine now then it should keep working fine, right? I also figure if it does fail in the future we can just fix it in a new wally version. Last I checked it's fine if one of the type exports in a module is broken.

> We could probably write a basic lexer and use that,

Which I'm all for if it's not too complex or if the complexity is worth it!

> though I'm not sure what it'd look like in practice.

and neither do I. 😞 It would be cool to see this explored.

> With regards to Rojo's project format I don't see it changing significantly any time soon. We should be able parse it and read through it with no difficulties, especially since we only really care about what the module turns into when it's require.

I agree! Also, if the project file does change significantly, we can _choose_ to support the change. "Explicitly not supporting future changes and not supporting new non-Rojo tools" is more about setting expectations than it is about what work we'll actually be doing. I don't suspect this to be an actual problem.
